besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/TrieLogHelper.java

```diff
@@ -0,0 +1,241 @@
+            MINIMUM_BONSAI_TRIE_LOG_RETENTION_THRESHOLD));
+    checkArgument(
+        config.getUnstable().getBonsaiTrieLogPruningLimit() > 0,
+        String.format(
+            "--Xbonsai-trie-log-pruning-limit=%d must be greater than 0",
+            config.getUnstable().getBonsaiTrieLogPruningLimit()));
+    checkArgument(
+        config.getUnstable().getBonsaiTrieLogPruningLimit()
+            > config.getUnstable().getBonsaiTrieLogRetentionThreshold(),
+        String.format(
+            "--Xbonsai-trie-log-pruning-limit=%d must greater than --Xbonsai-trie-log-retention-threshold=%d",
+            config.getUnstable().getBonsaiTrieLogPruningLimit(),
+            config.getUnstable().getBonsaiTrieLogRetentionThreshold()));
+  }
+
+  static TrieLogCount getCount(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final int limit,
+      final Blockchain blockchain) {
+    final AtomicInteger total = new AtomicInteger();
+    final AtomicInteger canonicalCount = new AtomicInteger();
+    final AtomicInteger forkCount = new AtomicInteger();
+    final AtomicInteger orphanCount = new AtomicInteger();
+    rootWorldStateStorage
+        .streamTrieLogKeys(limit)
+        .map(Bytes32::wrap)
+        .map(Hash::wrap)
+        .forEach(
+            hash -> {
+              total.getAndIncrement();
+              blockchain
+                  .getBlockHeader(hash)
+                  .ifPresentOrElse(
+                      (header) -> {
+                        long number = header.getNumber();
+                        final Optional<BlockHeader> headerByNumber =
+                            blockchain.getBlockHeader(number);
+                        if (headerByNumber.isPresent()
+                            && headerByNumber.get().getHash().equals(hash)) {
+                          canonicalCount.getAndIncrement();
+                        } else {
+                          forkCount.getAndIncrement();
+                        }
+                      },
+                      orphanCount::getAndIncrement);
+            });
+
+    return new TrieLogCount(total.get(), canonicalCount.get(), forkCount.get(), orphanCount.get());
+  }
+
+  private static void saveTrieLogsInFile(final Map<byte[], byte[]> trieLogs) throws IOException {
+
+    File file = new File(trieLogFile);
+
+    try (BufferedWriter bf = new BufferedWriter(new FileWriter(file, StandardCharsets.UTF_8))) {
+      for (Map.Entry<byte[], byte[]> entry : trieLogs.entrySet()) {
+        bf.write(Bytes.of(entry.getKey()) + ":" + Base64.toBase64String(entry.getValue()));
+        bf.newLine();
+      }
+      bf.flush();
+    } catch (IOException e) {
+      LOG.error(e.getMessage());
+      throw e;
+    }
+  }
+
+  private static IdentityHashMap<byte[], byte[]> readTrieLogsFromFile() throws IOException {
+
+    File file = new File(trieLogFile);
+    IdentityHashMap<byte[], byte[]> trieLogs = new IdentityHashMap<>();
```

Comment:
trielogs can vary in size, and the retention count is user configurable.  We could easily blow up the heap with bad settings.  We might want to add some sanity check on the file size here, and perhaps also in trielog streaming/save to file.

Reply:
Agree, users could eventually set the `logs-to-retain` to a high number and this could blow up. Is there real scenario where we should handle a high number having in mind that rolling back too many layer is not something feasible in bonsai (afaik)?

Reply:
Trielogs can be the basis of a specific worldstate for debug_ and trace_ endpoints.  But really if that were the case I imagine the node would be an archive node.  I think streaming to and from disk would likely just be more bulletproof.  

Not blocking feedback, just seems a robust mechanism

Reply:
I've changed this so we process it in batches now so it should be a safer mechanism.